### PR TITLE
Added UserBan

### LINKED ISSUE #145: Save if user is banned, so task will hide except user sends special flag
# Description

# Suggestion

### PR DESCRIPTION
**Pull request by msentwickler**
## Related Issue
https://github.com/E-Edu/task/issues/145

## Type of change

- [ ] Bug fix
- [x] Feature request
- [ ] Documenting code

## Checklists

- [x] I referenced the according Issue
  - [ ] I created an Issue
  - [x] I put a link in the Section [Related Issue](#related-issue)
- [ ] I tested my code with multiple browsers
- [x] My commit messages are good and understandable

### CODE REVIEW

File: src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+@AllArgsConstructor
+@Entity
+@Table
+public class UserBanModel {
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	private int bannedId;
+
+	@OneToOne
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java]: Why?

Comment (PAException) [reply src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java]: Delete that

File: src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java [hunk 2]
```diff
@@ -11,6 +14,8 @@
 @Service
 public class UserBanSubscriber extends AbstractObjectSubscriber<UserBanDTO> {
 
+	@Autowired private UserBanRepository userBanRepository;
+
 	@Getter private final Set<UUID> bannedUserIds = new HashSet<>();
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java]: Delete that

File: src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java [hunk 3]
```diff
@@ -20,6 +25,13 @@ public UserBanSubscriber() {
 	@Override
 	public void receive(UserBanDTO msg) {
 		this.getBannedUserIds().add(msg.getUserId());
+		userBanRepository.save(new UserBanModel(
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java]: in one line

File: src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java [hunk 4]
```diff
@@ -20,6 +25,13 @@ public UserBanSubscriber() {
 	@Override
 	public void receive(UserBanDTO msg) {
 		this.getBannedUserIds().add(msg.getUserId());
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java]: delete that

**Review by PAException — CHANGES_REQUESTED** [review]

**Comment by PAException** [comment]
Please also check now if you get tasks if the user is not banned

**Comment by msentwickler** [comment]
It's working.

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 5]
```diff
@@ -89,12 +91,16 @@
...
 
-	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey) {
+	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey, boolean showBanned) {
 		ControllerResult<LectureModel> lectureResult = this.lectureController
 			.getLectureByNameKey(lectureNameKey);
 		if (lectureResult.isResultNotPresent()) return ControllerResult.of(Error.NOT_FOUND, Constants.Lecture.NAME_KEY);
 
-		return ControllerResult.of(this.taskRepository.getAllTasksByLectureId(lectureResult.getResult()));
+		List<TaskModel> tasks = this.taskRepository.getAllTasksByLectureId(lectureResult.getResult());
+		tasks.removeIf(tm -> userBanRepository.existsByUserId(UUID.randomUUID()) || userBanRepository.existsByUserId(UUID.randomUUID()) && !showBanned);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: Why two times

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/TaskEndpoint.java [hunk 6]
```diff
@@ -36,8 +35,8 @@ public Object createTask(@RequestBody @Valid CreateTaskRequestDTO dto) {
 	}
 
 	@GetMapping("/lecture/{lectureNameKey}/task")
-	public Object getAllTasksFromLecture(@PathVariable @NotNull @NotEmpty @NotBlank String lectureNameKey) {
-		return this.taskController.getTasksFromLecture(lectureNameKey).getHttpResponse();
+	public Object getAllTasksFromLecture(@PathVariable @NotBlank String lectureNameKey, @RequestParam boolean showBanned) {
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/TaskEndpoint.java]: [Make RequestParam not required](https://www.baeldung.com/spring-request-param#making-an-optional-request-parameter)

**Review by PAException — CHANGES_REQUESTED** [review]
Please also check if user is banned on task/next

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 7]
```diff
@@ -68,7 +70,7 @@
 		if (lecture == null) return ControllerResult.of(Error.NOT_FOUND, Constants.Lecture.NAME_KEY);
 
 		List<TaskModel> tasks = this.taskRepository.getAllTasksByLectureId(lecture);
-		tasks.removeIf(tm -> finishedTaskIds.contains(tm.getTaskId())); //TODO: Use SQL query
+		tasks.removeIf(tm -> finishedTaskIds.contains(tm.getTaskId()) || userBanRepository.existsByUserId(UUID.randomUUID())); //TODO: Use SQL query
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: Use Constants.UserId.TEST_UUID

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 8]
```diff
@@ -68,7 +70,7 @@
...
 		if (tasks.isEmpty()) return ControllerResult.of(Error.NO_CONTENT, NAME_KEY);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: return NOT_FOUND

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 9]
```diff
@@ -101,12 +103,16 @@
...
 
-	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey) {
+	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey, boolean showBanned) {
 		ControllerResult<LectureModel> lectureResult = this.lectureController
 			.getLectureByNameKey(lectureNameKey);
 		if (lectureResult.isResultNotPresent()) return ControllerResult.of(Error.NOT_FOUND, Constants.Lecture.NAME_KEY);
 
-		return ControllerResult.of(this.taskRepository.getAllTasksByLectureId(lectureResult.getResult()));
+		List<TaskModel> tasks = this.taskRepository.getAllTasksByLectureId(lectureResult.getResult());
+		tasks.removeIf(tm -> userBanRepository.existsByUserId(UUID.randomUUID()) && !showBanned);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: Use Constants.UserId.TEST_UUID

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 10]
```diff
@@ -101,12 +103,16 @@
 		return ControllerResult.of(this.taskRepository.findAll());
 	}
 
-	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey) {
+	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey, boolean showBanned) {
 		ControllerResult<LectureModel> lectureResult = this.lectureController
 			.getLectureByNameKey(lectureNameKey);
 		if (lectureResult.isResultNotPresent()) return ControllerResult.of(Error.NOT_FOUND, Constants.Lecture.NAME_KEY);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: Use ControllerResult.ret

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java (hunk 9 above)
Comment (PAException) [reply src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: rename to task (tm)

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 11]
```diff
@@ -101,12 +103,16 @@
...
 
-		return ControllerResult.of(this.taskRepository.getAllTasksByLectureId(lectureResult.getResult()));
+		List<TaskModel> tasks = this.taskRepository.getAllTasksByLectureId(lectureResult.getResult());
+		tasks.removeIf(tm -> userBanRepository.existsByUserId(UUID.randomUUID()) && !showBanned);
+		if (tasks.isEmpty()) return ControllerResult.of(Error.NO_CONTENT, NAME_KEY);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: Return NOT_FOUND

File: src/main/java/de/themorpheus/edu/taskservice/database/repository/UserBanRepository.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package de.themorpheus.edu.taskservice.database.repository;
+
+import de.themorpheus.edu.taskservice.database.model.UserBanModel;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/database/repository/UserBanRepository.java]: ...

File: src/main/java/de/themorpheus/edu/taskservice/pubsub/dto/UserBanDTO.java [hunk 13]
```diff
@@ -4,15 +4,17 @@
...
 
 @Data
 @AllArgsConstructor
 @NoArgsConstructor
 @Component
-@PubSubDTO("user.ban")
+@PubSubDTO("report.user.ban")
 public class UserBanDTO {
 
+	@Column(columnDefinition = "BINARY(16)", nullable = false)
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/pubsub/dto/UserBanDTO.java]: Its not a Model in a db remove column annotation

File: src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java [hunk 14]
```diff
@@ -1,25 +1,36 @@
...
 	@Override
 	public void receive(UserBanDTO msg) {
-		this.getBannedUserIds().add(msg.getUserId());
+		userBanRepository.save(new UserBanModel(-1, msg.getUserId()));
 	}
 
+
+	@Override
+	public void deleteOrMaskUserData(UUID userId) {
+		userBanRepository.deleteAllByUserId(userId);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java]: use this.

File: src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java [hunk 15]
```diff
@@ -1,25 +1,36 @@
...
+	}
+
+	@Override
+	public Object getUserData(UUID userId) {
+		return userBanRepository.findAllByUserId(userId);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java]: use this.

File: src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java [hunk 16]
```diff
@@ -1,25 +1,36 @@
...
+	@Autowired private UserBanRepository userBanRepository;
 
 	public UserBanSubscriber() {
 		super(UserBanDTO.class);
 	}
 
 	@Override
 	public void receive(UserBanDTO msg) {
-		this.getBannedUserIds().add(msg.getUserId());
+		userBanRepository.save(new UserBanModel(-1, msg.getUserId()));
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java]: use this.

File: src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table
+public class UserBanModel {
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	private int bannedId;
+
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java:24]: Add @ NotNull

File: src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table
+public class UserBanModel {
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	private int bannedId;
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java]: change to userBanId

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java (hunk 7 above)
Comment (PAException) [reply src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: use `this.`

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java (hunk 9 above)
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: use this.

**Review by PAException — CHANGES_REQUESTED** [review]

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java (hunk 10 above)
Comment (msentwickler) [reply src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: Why in this line?
Did you mean in the next line?

File: src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java (hunk 17 above)
Comment (msentwickler) [reply src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java:24]: Do I need this if it's auto generated?

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java (hunk 10 above)
Comment (PAException) [reply src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: no use ControllerResult.ret(lectureResult)

File: src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java (hunk 17 above)
Comment (PAException) [reply src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java:24]: the userId
and add column label

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 19]
```diff
@@ -68,8 +70,8 @@
 		if (lecture == null) return ControllerResult.of(Error.NOT_FOUND, Constants.Lecture.NAME_KEY);
 
 		List<TaskModel> tasks = this.taskRepository.getAllTasksByLectureId(lecture);
-		tasks.removeIf(tm -> finishedTaskIds.contains(tm.getTaskId())); //TODO: Use SQL query
-		if (tasks.isEmpty()) return ControllerResult.of(Error.NO_CONTENT, NAME_KEY);
+		tasks.removeIf(tm -> finishedTaskIds.contains(tm.getTaskId()) || this.userBanRepository.existsByUserId(Constants.UserId.TEST_UUID)); //TODO: Use SQL query
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: Use task

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 20]
```diff
@@ -101,12 +103,16 @@
...
 
-	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey) {
+	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey, boolean showBanned) {
 		ControllerResult<LectureModel> lectureResult = this.lectureController
 			.getLectureByNameKey(lectureNameKey);
-		if (lectureResult.isResultNotPresent()) return ControllerResult.of(Error.NOT_FOUND, Constants.Lecture.NAME_KEY);
+		if (lectureResult.isResultNotPresent()) return ControllerResult.ret(lectureResult);
+
+		List<TaskModel> task = this.taskRepository.getAllTasksByLectureId(lectureResult.getResult());
+		task.removeIf(tm -> this.userBanRepository.existsByUserId(Constants.UserId.TEST_UUID) && !showBanned);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java]: Use task

File: src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java (hunk 17 above)
Comment (PAException) [reply src/main/java/de/themorpheus/edu/taskservice/database/model/UserBanModel.java:24]: NotNull and don't needed bc of nullable=false

**Review by PAException — CHANGES_REQUESTED** [review]

File: src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java [hunk 21]
```diff
@@ -1,25 +1,36 @@
...
+
+	@Override
+	public void deleteOrMaskUserData(UUID userId) {
+		this.userBanRepository.deleteAllByUserId(userId);
+	}
+
+	@Override
+	public Object getUserData(UUID userId) {
+		return this.userBanRepository.findAllByUserId(userId);
+	}
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/pubsub/sub/UserBanSubscriber.java:35]: empty line

**Review by PAException — CHANGES_REQUESTED** [review]

**Review by PAException — APPROVED** [review]

**Pull request merged**